Network/Wai/Middleware/Static.hs

```diff
@@ -56,6 +59,26 @@ import qualified System.FilePath as FP
 newtype Policy = Policy { tryPolicy :: String -> Maybe String -- ^ Run a policy
                         }
 
+-- | Options for 'staticWithOptions' 'Middleware'.
+data Options = Options { cacheContainer :: CacheContainer -- ^ Cache container to use
+                       , mimeTypes :: FilePath -> MimeType -- ^ Compute MimeType from file name
+                       }
+
+-- | Default options.
+--
+-- 'cacheContainer' = 'CacheContainerEmpty' -- no caching
+-- 'mimeTypes' = 'getMimeTypes' -- use 'defaultMimeLookup' from 'Network.Mime'
+defaultOptions :: Options
+defaultOptions = Options { cacheContainer = CacheContainerEmpty, mimeTypes = getMimeType }
+
+-- | Update cacheContainer of options.
+setCacheContainer :: CacheContainer -> Options -> Options
```

Comment:
Is this function necessary? You already export the `cacheContainer` record selector, so updating it should simply be a matter of doing `options { cacheContainer = myCacheContainer }`, right?

Reply:
I'm exporting `Options` abstractly, so the fact that it is a record is not exposed. This is part of the "heavyweight" aspect I mentioned before and is probably overkill.

Reply:
You can still use record accessors and updates even if the `Options` constructor is not exposed:

```
foo, bar :: CacheContainer
foo = cacheContainer defaultOptions
bar = cacheContainer (defaultOptions { cacheContainer = foo })
```

Even if we use the heavyweight approach (which I think is a good idea), we shouldn't need to define these setter functions separately.

Reply:
It might help to advertise that these are record selectors as well. [`Data.Aeson`](http://hackage.haskell.org/package/aeson-1.4.5.0/docs/Data-Aeson.html#t:Options) does this, for instance. ("Options can be set using record syntax on `defaultOptions` with the fields below.")

Reply:
My original intent was not to have the fields be viewed or usable as fields as that partially defeats the point of the approach, but instead I've just changed the code to not treat `Options` as an abstract type and removed the `set*` functions.

Reply:
I'm not sure why that defeats the point of the approach myself. The way I see it, exporting `Options` abstractly alongside its record selectors is the best of both worlds, since:

1. Adding new fields to `Options` in the future won't require a major version bump (since the change to the type of the `Options` data constructor won't be observable), and
2. You can use record syntax to create/access/update `Options`, which is syntactically more convenient than having to define setter functions.

This code is _almost_ there. We just need not to export the `Options` data constructor. If you can do that, I'll merge.

Reply:
It *partially* defeats the point because you may want an option to stop being a field. For example, say you have a Boolean option that specifies whether an uploaded file should overwrite an existing file or fail, but then you later decide there should be a third possibility of allowing the upload but renaming the file. With a completely opaque options type, you could just make the setter/accessor for the old option be aliases for setting/testing the new option and add a setter/accessor for the new option. With fields, the alternatives are to change/remove the old option, which is not a backwards compatible change, or keep the old field and add a new field plus some potentially non-obvious logic to combine the possibly conflicting options. Essentially, Haskell lacks a notion of "virtual" fields.

As I said, though, this is probably overkill. I don't have a strong opinion about this.